components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,478 @@
+type uln2003 struct {
+	theBoard           board.Board
+	rotationPerMinute  float64
+	ticksPerRotation   int
+	in1, in2, in3, in4 board.GPIOPin
+	logger             golog.Logger
+	motorName          string
+
+	// state
+	lock  sync.Mutex
+	opMgr operation.SingleOperationManager
+
+	stepPosition         int64
+	threadStarted        bool
+	targetStepPosition   int64
+	targetStepsPerSecond int64
+	generic.Unimplemented
+}
+
+// validate if this config is valid.
+func (m *uln2003) Validate() error {
+	if m.theBoard == nil {
+		return errors.Errorf("need a board for motor (%s)", m.motorName)
+	}
+
+	if m.rotationPerMinute == 0 {
+		return errors.Errorf("need to set 'rotation_per_minute' for motor (%s)", m.motorName)
+	}
+
+	if m.in1 == nil {
+		return errors.Errorf("need a 'In1' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in2 == nil {
+		return errors.Errorf("need a 'In2' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in3 == nil {
+		return errors.Errorf("need a 'In3' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in4 == nil {
+		return errors.Errorf("need a 'In4' pin for motor (%s)", m.motorName)
+	}
+
+	return nil
+}
+
+// SetPower sets the percentage of power the motor should employ between 0-1.
+func (m *uln2003) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	if math.Abs(powerPct) <= .0001 {
+		m.stop()
+		return nil
+	}
+	return errors.Errorf("doesn't support raw power mode in motor (%s)", m.motorName)
+}
+
+func (m *uln2003) startThread(ctx context.Context) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.threadStarted {
+		return
+	}
+
+	m.threadStarted = true
+	go m.doRun(ctx)
+}
+
+// setStepperDelay calculates the wait time between each step taken.
+// The minimum value is set to 0.002s, anything less then this can potentially damage the gears.
+func (m *uln2003) setStepperDelay(rpm float64) float64 {
+	stepperDelay := 1 / ((rpm * 8) / 60)
+	return math.Max(stepperDelay, minStepDelay)
+}
+
+func (m *uln2003) doRun(ctx context.Context) {
+	for {
+		sleep, err := m.doCycle(ctx)
+		if err != nil {
+			m.logger.Info("error in uln2003 %w", err)
+		}
+
+		if !utils.SelectContextOrWait(ctx, sleep) {
+			return
+		}
+	}
+}
+
+func (m *uln2003) doCycle(ctx context.Context) (time.Duration, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.stepPosition == m.targetStepPosition {
+		err := m.enable(ctx, false)
+		return 5 * time.Millisecond, err
+	}
+
+	err := m.doStep(ctx, m.stepPosition < m.targetStepPosition, m.rotationPerMinute)
+	if err != nil {
+		return time.Second, fmt.Errorf("error stepping %w", err)
+	}
+	k := time.Duration(int64(time.Microsecond*1000*1000) / int64(math.Abs(float64(m.targetStepsPerSecond))))
+	return k, nil
+}
+
+// doStep has to be locked to call.
+// Depending on the direction, doStep will either treverse the stepSequence array in ascending
+// or descending order.
+func (m *uln2003) doStep(ctx context.Context, forward bool, rpm float64) error {
+	if forward {
```

Comment:
we can reduce the amount of code in this function by having a local variable that copies `stepSequence` is `forward` is true and makes a reverse copy otherwise. Then you could accomplish this all with a single for-loop